app/models/course_user.rb

```diff
@@ -176,6 +186,22 @@ def self.user?(user)
     all.exists?(user: user)
   end
 
+  # Soft delete the course user.
+  def destroy
```

Comment:
Soft-delete of course-user with a lot of activities still take >5 minutes. Your PR does not resolve the root issue. Please revise.

Reply:
The only way is verify by similar data volumn on production. We can keep monitoring it on staging to see.
In theory, the soft-delete won't effect to the index and fk-constrains, so it would be much faster than hard-delete.

Reply:
I already verified the slow deletion using production-grade data, the problem is that there are many individual associated updates deletes involved.  

I suggest we simply do soft-delete on course-user and do not touch associated models.  
Subsequent calls to query data should ensure scope of course-user does not include soft-deleted course-user. `discard` gem might be better for this.

Reply:
Spend time researching about discard. Might apply it later then.